File: go/fixture/go-mini/internal/repository/file_repo.go
package repository

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"slices"
	"strings"
	"sync"

	"example.com/go-mini/internal/models"
)

// FileRepo is a repo backed by a file.
type FileRepo struct {
	path string
	mu   sync.Mutex
}

// NewFileRepo creates a new FileRepo.
func NewFileRepo(path string) (*FileRepo, error) {
	if path == "" {
		return nil, errors.New("file repo: empty path")
	}
	return &FileRepo{path: path}, nil
}

// Get gets a device.
func (r *FileRepo) Get(ctx context.Context, tenant, id string) (models.Device, error) {
	if err := ctx.Err(); err != nil {
		return models.Device{}, fmt.Errorf("file repo: get: %w", err)
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	devices, err := r.load()
	if err != nil {
		return models.Device{}, err
	}
	d, ok := devices[key(tenant, id)]
	if !ok {
		return models.Device{}, ErrNotFound
	}
	return d, nil
}

// Save saves a device.
func (r *FileRepo) Save(ctx context.Context, d models.Device) error {
	if err := ctx.Err(); err != nil {
		return fmt.Errorf("file repo: save: %w", err)
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	devices, err := r.load()
	if err != nil {
		return err
	}
	devices[key(d.Tenant, d.ID)] = d
	return r.store(devices)
}

// List lists all devices.
func (r *FileRepo) List(ctx context.Context) ([]models.Device, error) {
	if err := ctx.Err(); err != nil {
		return nil, fmt.Errorf("file repo: list: %w", err)
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	devices, err := r.load()
	if err != nil {
		return nil, err
	}
	return sorted(devices), nil
}

func (r *FileRepo) load() (map[string]models.Device, error) {
	data, err := os.ReadFile(r.path)
	if errors.Is(err, os.ErrNotExist) {
		return map[string]models.Device{}, nil
	}
	if err != nil {
		return nil, fmt.Errorf("file repo: read %s: %w", r.path, err)
	}
	devices := map[string]models.Device{}
	if err := json.Unmarshal(data, &devices); err != nil {
		return nil, fmt.Errorf("file repo: decode %s: %w", r.path, err)
	}
	return devices, nil
}

func (r *FileRepo) store(devices map[string]models.Device) error {
	data, err := json.MarshalIndent(devices, "", "  ")
	if err != nil {
		return fmt.Errorf("file repo: encode: %w", err)
	}
	if err := os.WriteFile(r.path, data, 0o600); err != nil {
		return fmt.Errorf("file repo: write %s: %w", r.path, err)
	}
	return nil
}

func sorted(devices map[string]models.Device) []models.Device {
	out := make([]models.Device, 0, len(devices))
	for _, d := range devices {
		out = append(out, d)
	}
	slices.SortFunc(out, func(a, b models.Device) int {
		return strings.Compare(key(a.Tenant, a.ID), key(b.Tenant, b.ID))
	})
	return out
}

File: py/fixture/py-mini/internal/repository/file_repo.py
"""A device store backed by one JSON file."""

import json
import threading
from datetime import datetime
from typing import Any

from internal.models.device import Device
from internal.repository.store import NotFoundError, _key


class FileRepo:
    """A repo backed by a file."""

    def __init__(self, path: str) -> None:
        """Create a new FileRepo."""
        if not path:
            raise ValueError("file repo: empty path")
        self.path = path
        self._guard = threading.Lock()

    def get(self, tenant: str, device_id: str) -> Device:
        """Get a device."""
        with self._guard:
            devices = self._load()
        d = devices.get(_key(tenant, device_id))
        if d is None:
            raise NotFoundError(f"device {tenant}/{device_id} not found")
        return d

    def save(self, d: Device) -> None:
        """Save a device."""
        with self._guard:
            devices = self._load()
            devices[_key(d.tenant, d.id)] = d
            self._store(devices)

    def list_devices(self) -> list[Device]:
        """List all devices."""
        with self._guard:
            devices = self._load()
        if not devices:
            return []
        return _sorted(devices)

    def _load(self) -> dict[str, Device]:
        try:
            with open(self.path, encoding="utf-8") as f:
                data = json.load(f)
        except FileNotFoundError:
            return {}
        except (OSError, ValueError) as err:
            raise OSError(f"file repo: read {self.path}: {err}") from err
        if not isinstance(data, dict):
            raise OSError(f"file repo: decode {self.path}: not an object")
        return {k: _from_json(v) for k, v in data.items()}

    def _store(self, devices: dict[str, Device]) -> None:
        data = {k: _to_json(d) for k, d in devices.items()}
        try:
            with open(self.path, "w", encoding="utf-8") as f:
                json.dump(data, f, indent=2)
        except OSError as err:
            raise OSError(f"file repo: write {self.path}: {err}") from err


def _to_json(d: Device) -> dict[str, Any]:
    return {
        "id": d.id,
        "tenant": d.tenant,
        "status": d.status,
        "version": d.version,
        "tags": list(d.tags),
        "last_seen": d.last_seen.isoformat(),
    }


def _from_json(v: dict[str, Any]) -> Device:
    return Device(
        id=v["id"],
        tenant=v["tenant"],
        status=v["status"],
        version=v.get("version", ""),
        tags=list(v.get("tags") or []),
        last_seen=datetime.fromisoformat(v["last_seen"]),
    )


def _sorted(devices: dict[str, Device]) -> list[Device]:
    return [devices[k] for k in sorted(devices)]
